### PR TITLE
Add LspPeekDefinition

### LINKED ISSUE #411: Add "peek definition"
I'd be nice to add a `LspPeekDefinition` (insert bike-shedding here, name taken from VS Code that offers this) that works like `LspDefinition`, but instead of jumping to the location in a new buffer, only shows the location in a preview window (like `LspHover` uses; bonus points for floating window when that is implemented).

### PR DESCRIPTION
**Pull request by thomasfaingnaert**
Adds `:LspPeekDefinition`, which displays the definition in the preview window, similar to VSCode's peek definition.

Fixes #411

### CODE REVIEW

**Comment by prabirshrestha** [comment]
Can you make use of the changes that came in this PR which just got merged. https://github.com/prabirshrestha/vim-lsp/pull/395

**Comment by thomasfaingnaert** [comment]
I pushed a basic version that works using:
- preview:
preview

- Vim popups:
vim popup

- Neovim floats:
neovim float

**Comment by prabirshrestha** [comment]
@jerdna-regeiz could you have a look at this PR first.

File: ftplugin/lsp-hover.vim [hunk 1]
```diff
@@ -7,13 +7,13 @@ else
   setlocal previewwindow buftype=nofile bufhidden=wipe noswapfile nobuflisted
 endif
 setlocal nocursorline nofoldenable
+setlocal nonumber norelativenumber
```
Comment (jerdna-regeiz) [inline ftplugin/lsp-hover.vim]: why not just put it in the line above?
```suggestion
setlocal nocursorline nofoldenable nonumber norelativenumber
```

File: plugin/lsp.vim [hunk 2]
```diff
@@ -63,7 +64,8 @@ command! LspPreviousReference call lsp#ui#vim#references#jump(-1)
 
 nnoremap <plug>(lsp-code-action) :<c-u>call lsp#ui#vim#code_action()<cr>
 nnoremap <plug>(lsp-declaration) :<c-u>call lsp#ui#vim#declaration()<cr>
-nnoremap <plug>(lsp-definition) :<c-u>call lsp#ui#vim#definition()<cr>
+nnoremap <plug>(lsp-definition) :<c-u>call lsp#ui#vim#definition(0)<cr>
+nnoremap <plug>(lsp-peek-definition) :<c-u>call lsp#ui#vim#definition(1)<cr>
```
Comment (jerdna-regeiz) [inline plugin/lsp.vim]: Could we use `v:true` and `v:false` to express more clearly that it is boolean? Or does it really require an int somewhere I missed? (same goes for the command)

**Comment by jerdna-regeiz** [comment]
When `win_execute` is implemented in nvim we should use that instead of switching focus multiple times (maybe we could even manage without switching at all). But for now I have found no other solution as well :( Variables could be set with `nvim_win_set_{var,option}`, but I have no alternative for calling `zz` and the like for adjustment.

File: plugin/lsp.vim (hunk 2 above)
Comment (prabirshrestha) [reply plugin/lsp.vim]: Old version of neovim doesn’t support v:true and it is buggy too.

Comment (jerdna-regeiz) [reply plugin/lsp.vim]: okay, didn't know that

File: ftplugin/lsp-hover.vim (hunk 1 above)
Comment (thomasfaingnaert) [reply ftplugin/lsp-hover.vim]: That's indeed better

File: autoload/lsp/ui/vim/output.vim [hunk 3]
```diff
@@ -213,9 +272,28 @@ function! lsp#ui#vim#output#preview(data) abort
       if has('nvim')
         call s:adjust_float_placement(l:bufferlines, l:maxwidth)
         call s:add_float_closing_hooks()
+
+        if has_key(a:options, 'cursor')
+          let s:suppress_close = v:true
```
[the remark this answers is not in the document]
Comment (thomasfaingnaert) [reply autoload/lsp/ui/vim/output.vim]: I completely missed that, thanks!

**Comment by thomasfaingnaert** [comment]
@jerdna-regeiz Thanks for reviewing my PR! Regarding my TODO, I tried `win_execute` with `normal! zz` and friends, and that doesn't seem to work. Do you know a way to do the alignment in Vim? At the moment, only setting the `firstline` seems to work, but that's kind of annoying to use because you would have to know the window height in advance, and calculate the position based on that. (At least for center and bottom)

**Comment by jerdna-regeiz** [comment]
@thomasfaingnaert Have you tried to set the option afterwards with `popup_setoptions` like when the resize is done for nvim, because the size/content is known then? 
We should up the version check to patch 8.1.1561 for popup_setoptions to work.

**Comment by thomasfaingnaert** [comment]
There's also quite a bit of complex conditionals, which I'm not a big fan of. Perhaps we should set 3 boolean script variables `s:use_preview`, `s:use_vim_popup`, `s:use_neovim_float` instead of always having to type things like 
```vim
if s:supports_floating && g:lsp_preview_float && !has('nvim')
```

**Comment by jerdna-regeiz** [comment]
> There's also quite a bit of complex conditionals, which I'm not a big fan of. Perhaps we should set 3 boolean script variables `s:use_preview`, `s:use_vim_popup`, `s:use_neovim_float` instead of always having to type things like

I agree, but it could be done in a separate PR to not mix a new feature with refactoring.

**Comment by thomasfaingnaert** [comment]
Tried fixing Vim alignment of the popup window in ccaf0f95b4cc9d9835b272a16c5ca4e27d7e0cf2, but I'm facing one more issue: `l:height` is 1, as if the popup still contains `...`, even though the content should already be set by then, right?

**Comment by thomasfaingnaert** [comment]
Seems like using a timer and specifying the width and height explicitly works.

**Comment by thomasfaingnaert** [comment]
Added Peek commands for implementation, type definition and declaration. I have not been able to test these, because I don't have a language server that supports these, but they should work the same as LspPeekDefinition.

As for LspReferences, I'm not sure if it would make sense to use the popup window for those, because you have to interact with the quickfix list to go through the list of references?

**Comment by clason** [comment]
I think it's fine as is -- if somebody comes up with a use case for `PeekReferences` (I don't see one either), that could be added in a future PR. LGTM!

**Comment by prabirshrestha** [comment]
@thomasfaingnaert  This has merge conflicts. Can you fix it?

**Comment by thomasfaingnaert** [comment]
@prabirshrestha Fixed merge conflicts.
I'm not sure what changed in the meantime, but Vim's floating window isn't positioned well anymore: it seems to place the preview above the cursor even when there are only 2 lines available above:
image
I fixed this in 18fcb436b1ef3fc68437133bc0628e2e60d4fa52 by forcing the popup above/below the cursor depending on if the cursor is in the first or second half of the screen.
Should I change this, or do we leave it as is?

**Comment by prabirshrestha** [comment]
@thomasfaingnaert Current behavior is good.

I had also sent you an email but haven't got any reply. Can you send me an email. You can find my email at https://github.com/prabirshrestha

**Pull request merged**